system/datatypes/mailinglist.php

```diff
@@ -458,16 +539,38 @@ private function addressesView(HyphaRequest $request) {
 		$main = $this->html->find('#main');
 		$main->appendChild($table);
 		$table->addClass('section');
-		$table->addHeaderRow()->addCells(['', __('addresses'), __('status')]);
+		$table->addHeaderRow()->addCells(['', __('addresses'), __('status'), __('action')]);
 		/** @var HyphaDomElement[] $addresses */
 		$xpath = './/' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_STATUS . '="' . self::ADDRESS_STATUS_CONFIRMED . '" or @' . self::FIELD_NAME_STATUS . '="' . self::ADDRESS_STATUS_PENDING . '"]';
 		$addresses = $this->findAddresses($xpath);
 		foreach ($addresses as $address) {
 			$status = $address->getAttribute(self::FIELD_NAME_STATUS);
 			$statusText = __('ml-address-status-' . $status);
-			$table->addRow()->addCells(['', $address->getAttribute(self::FIELD_NAME_EMAIL), $statusText]);
+			$actions = [];
+			if (isAdmin()) {
+				if ($status === self::ADDRESS_STATUS_PENDING) {
+					$confirmCode = $address->getAttribute(self::FIELD_NAME_CONFIRM_CODE);
+					if ($address->getAttribute(self::FIELD_NAME_REMINDED) != true) {
+						$remindPath = $this->substituteSpecial(self::PATH_REMIND_CODE, ['code' => $confirmCode]);
+						$actions[] = $this->makeActionButton(__('ml-send-reminder'), $remindPath);
+					} else {
+						$statusText .= ', ' . __('ml-reminded');
+					}
+					$deletePath = $this->substituteSpecial(self::PATH_DELETE_ADDRESS_CODE, ['code' => $confirmCode]);
+					$actions[] = $this->makeActionButton(__('delete'), $deletePath);
+				}
+				if ($status === self::ADDRESS_STATUS_CONFIRMED) {
+					$unsubscribeCode = $address->getAttribute(self::FIELD_NAME_UNSUBSCRIBE_CODE);
```

Comment:
Again, I would use an `isAdmin()` check in `unsubscribeAction()` rather than reusing the existing unsubscription code. Here, leaking the code is less of an issue, but checking `isAdmin()` in the action makes it more explicit that unsubscription is also allowed by admins.

Reply:
fixed

Reply:
Really? I do not see any change in `unsubscribeAction` and the url still contains the unsubscribe code?

Reply:
The unsubscribeAction is used for both admins and subscribers. So the `isAdmin()` isn't a good idea. I could make an extra action for the admins, but why, this one is already there?!

Reply:
So, yeah, I left it as it is.

Reply:
That's not really "fixed" :-)

I can imagine that the same action can be authenticated using `isAdmin()`, or failing that, require a code to be specified? Or two actions that call the same function to do the actual work?

As for why: it seems cleaner to me when the authentication code intended to be used for the recipient is not exposed elsewhere. Additionally, adding the isAdmin() check to `unsubscribeAction` makes it more clear to someone reading the code that unsubscription can be done by admins (iow, all access control appears in the same place, namely the action methods).

However, as I said, leaking the unsubscribe code is not as problematic as leaking the confirm code, so when I fail to convince you, leaving this as is is also an option for me :-)

Reply:
implemented new method.